### PR TITLE
Native-specific implementation.

### PR DESCRIPTION
**Pull request by lrhn**
Splits implementation of `Int64` into a native version and a web version, using conditional imports.

This PR is, so far, mainly a request for comments on the approach.

The problem with conditional imports (or exports in this case) is that the non-imported library still exists in the code, and defines a type with the same name, and the analyzer shows errors where the type names conflict, because both `Int64` implementations currently refer to the same `Int32` implementation, which in turn, according to the analyzer, refers back to the *default* `Int64` implementation only. The non-default implementation gets the wrong `Int64` when it calls `Int32.toInt64()`.

I could just make separate `Int32` implementations too. Then we only get the problem when accessing `NumX.toInt64()` (and now also `toInt32()`). 

An alternative approach defines a separate `Int64` interface, and makes the implementation classes separate. Then they only need to agree on the interface type.
The problem with doing that here, is that the `Int64` class is written in a way where it assumes the interface is not implemented in more than one way, all `Int64` arguments are assumed to provide access to the private members.
With a shared interface, arguments then requires casting each `Int64` argument to `_Int64Impl` in order to access the internal value.

Maybe it's possible to work around that, e.g, using `.toInt()` instead of `._value` in the native version, which should always give the same result. 
The web version can't do that, though, because it needs access to the internal `_h`, `_m` and `_l` fields. On the other hand, it might be able to elude the casts, so it won't cost extra.

WDYT? (Also, is it even worth doing?)

### CODE REVIEW

File: lib/src/utilities.dart [hunk 1]
```diff
@@ -70,3 +70,18 @@ int bitCount(int i) {
...
+
+void reverseList(List<Object?> values) {
+  var half = values.length ~/ 2;
+  var last = values.length - 1;
+  for (var i = 0; i < half; i++) {
+    var tmp = values[i];
+    var fromEnd = last - i;
+    values[i] = values[fromEnd];
+    values[fromEnd] = tmp;
+  }
```
Comment (rakudrama) [inline lib/src/utilities.dart:86]: I didn't notice where this is used.

I find it easier to understand the indexing when written like this:
```
int start = 0, end = values.length-1;
while (start < end) {
  final first = values[start];
  final last = values[end];
  values[start++] = last;
  values[end--] = first;
}
```
This also generalizes nicely for a reverse-subrange operation.

**Comment by kevmoo** [comment]
Analyzer is sad. Also: should we add a benchmark here BEFORE this change so we can do a before/after?

File: lib/src/int64.dart [hunk 2]
```diff
@@ -2,1128 +2,7 @@
...
-        return ZERO;
-      } else {
-        return _sub(b0, b1, b2, r0, r1, r2);
-      }
-    } else {
-      return _negate(r0, r1, r2);
-    }
-  }
-}
+// Use export below instead while working on the web implementation.
```
Comment (rakudrama) [inline lib/src/int64.dart:7]: It seems off that you would need this  - just work on the web implementation with a web compiler.

**Comment by rakudrama** [comment]
Is it worth doing? Slava mentioned that it would be 'nice' for native protobufs, but said that `Int64` did not show on the profiles. So perhaps it is not worth doing until that changes.

If the problem is that the analyzer is confused by a cycle with conditional imports, we should fix that.
Surely the `int32.dart` imports `int64.dart`, and hence the conditional `Int64` definition?

Perhaps make the conditional import be at a higher level, above the cycle
- `fixnum.dart` conditionally exports `fixnum_native.dart` or `fixnum_web.dart`.
-  or an internal library that implements both `Int32` and `Int64`, that share `intx.dart` that imports the platform specific library

If you do decide to implement an interface, I prefer a single implementation of `Foo` to be named `_Foo` rather than `_FooImpl`. 'Impl' is not a word.
I don't understand the problem with the implementation subtype. All the operations call `_promote`, which could exist in both platform libraries and return the private implementation type.

I have some other comments but I can't figure out how to comment on unchanged lines in a github code review .
---
I think we should consider deprecating `IntX` and `Int32`.
There are few uses of `Int32` (most grep-style matches are `ffi.Int32`, a different class for ffi annotations).

Many uses of `Int32` are work-arounds for missing basic functionality
```
final int _maxInt = Int32.MAX_VALUE.toInt();  // Missing handy constant
```

`Int32(v).toInt()` --> `v.toSigned(32)`: I suspect the author simply didn't find or understand `toSigned`.

`(Int32(a) * Int32(b)).toInt()`: it would be better to expose a multiply32 operation that can be easily implemented on the web via JavaScript's `Math.imul32`. 

I can't find a use of `Int32.numberOfLeadingZeros`/`Int32.numberOfTrailingZeros`, but this is a longstanding request on standard `int` values (https://github.com/dart-lang/sdk/issues/38346).

File: lib/src/utilities.dart (hunk 1 above)
Comment (lrhn) [reply lib/src/utilities.dart:86]: Good catch, I stopped using it again.

**Comment by lrhn** [comment]
> I think we should consider deprecating IntX and Int32.

YES! I was thinking that too, but didn't dare hope that it would be possible.
It would be much easier to make `Int64` focused without having to worry about being generalized as an `IntX`.

There seems to be lots of uses of `Int32` in internal code, but some are definitely the `dart:ffi` type, and the first many uses of `fixnum`'s type that I found were of the form `Int32.MAX_VALUE.toInt()`. Also saw some `Int32(expression).toInt()` which should just be using `.toSigned(32)`. 
Still, there are some uses left.

**Comment by natebosch** [comment]
> I could just make separate `Int32` implementations too. Then we only get the problem when accessing `NumX.toInt64()` (and now also `toInt32()`).

We could make separate `IntX` too, I don't like the duplication, but it might be better to live with the copy/paste than the other workarounds.

**Comment by lrhn** [comment]
Another thought is that most people don't need `Int64` with all the operations. They just need `Id64` which holds a 64 bit ID number, and has `==`/`hashCode`, `toString` and `parse`. 
Forcing these people to use `Int64` is overkill. 

We could add `Id64` and make `Int64` implement it, to make migration possible.

**Comment by rakudrama** [comment]
> Another thought is that most people don't need `Int64` with all the operations.

Protocol buffers are an important user of `Int64`. We don't really know all the uses of 64-bit integers in protocol buffers but it is safe to assume that all the 64-bit operations available to C++ will be used somewhere.

The binary protobuf serialization format needs quite a few 64-bit operations to encode the value in the byte stream (varint and zigzag encodings do comparison, shifting, masking etc). 64-bit values often encode 'micro-currencies', a representation scaled by 10<sup>6</sup>. This is subject to arithmetic for business logic and formatting.

**Comment by kevmoo** [comment]
Is this dead?

**Comment by lrhn** [comment]
Haven't had time for it recently. Let's say "yes" for now. I can always open it again if I find time.

**Pull request closed without merging**